### PR TITLE
Branch

### PR DESCRIPTION
**Pull request by nvuletic**

### CODE REVIEW

File: application/Application.cpp [hunk 1]
```diff
@@ -241,6 +365,7 @@ int main(int argc, char** argv)
 
     modbusBridge->start();
 
+    // This while loop, even if it seems not necessary, is actually necessary.
```
Comment (milanzlicic) [inline application/Application.cpp]: No useless comments

File: src/modbus/ModbusBridge.cpp [hunk 2]
```diff
@@ -172,6 +196,140 @@ void ModbusBridge::handleActuation(const std::string& /* deviceKey */, const std
...
+                                                 modbusRegisterMapping.getAddress(), uShortValue))
+        {
+            LOG(ERROR) << "ModbusBridge: Unable to write holding register value - Register address: "
+                       << modbusRegisterMapping.getAddress() << " Value: " << uShortValue;
+            modbusRegisterWatcher.setValid(false);
+        }
+    }
+    else if (modbusRegisterMapping.getDataType() == ModbusRegisterMapping::DataType::REAL32)
+    {
+        float floatValue = static_cast<float>(atoi(value.c_str()));
```
Comment (milanzlicic) [inline src/modbus/ModbusBridge.cpp]: Shouldn't this be atof?

File: src/modbus/ModbusBridge.cpp [hunk 3]
```diff
@@ -172,6 +196,140 @@ void ModbusBridge::handleActuation(const std::string& /* deviceKey */, const std
...
+                       << modbusRegisterMapping.getAddress();
+            modbusRegisterWatcher.setValid(false);
+        }
+    }
+}
+
+void ModbusBridge::handleConfigurationForCoil(const ModbusRegisterMapping& modbusRegisterMapping,
+                                              ModbusRegisterWatcher& modbusRegisterWatcher, const std::string& value)
+{
+    bool boolValue = (value == "true");
```
Comment (milanzlicic) [inline src/modbus/ModbusBridge.cpp]: Is this only ever "true" or "false", or can it be "0", "1" etc...?

File: src/modbus/ModbusBridge.cpp [hunk 4]
```diff
@@ -331,6 +489,159 @@ wolkabout::ActuatorStatus ModbusBridge::getActuatorStatusFromCoil(
...
+                modbusRegisterWatcher.setValid(false);
+                return ConfigurationItem(std::vector<std::string>(), "");
+            }
+            modbusRegisterWatcher.update(value);
+        }
+    }
+
+    std::string s;
+    std::stringstream stream(modbusRegisterWatcher.getValue());
+    while (getline(stream, s, ','))
```
Comment (milanzlicic) [inline src/modbus/ModbusBridge.cpp]: I think there is a split function in string utils

File: src/modbus/ModbusBridge.cpp [hunk 5]
```diff
@@ -437,9 +748,25 @@ void ModbusBridge::readAndReportModbusRegistersValues()
...
+                        }
+                        break;
+                    case ModbusRegisterMapping::MappingType::CONFIGURATION:
+                        if (m_onConfigurationChange)
+                        {
+                            m_onConfigurationChange();
+                        }
+                        break;
+                    default:
+                        LOG(ERROR) << "This mapping wasn\'t declared properly!";
```
Comment (milanzlicic) [inline src/modbus/ModbusBridge.cpp]: What mapping, add more descriptive error

File: src/modbus/ModbusRegisterMapping.h [hunk 6]
```diff
@@ -34,6 +38,15 @@ class ModbusRegisterMapping
         HOLDING_REGISTER_ACTUATOR
     };
 
+    enum class MappingType
+    {
+        DEFAULT,
```
[the remark this answers is not in the document]
Comment (nvuletic) [reply src/modbus/ModbusRegisterMapping.h]: DEFAULT just invokes the behavior of past versions of module. That is assigned to mapping if they don't specify it, and in code, it calls the code that was used before to decide whether or not it was a Sensor/Actuator

File: src/modbus/ModbusRegisterMapping.cpp [hunk 7]
```diff
@@ -102,29 +154,133 @@ std::vector<wolkabout::ModbusRegisterMapping> ModbusRegisterMappingFactory::from
...
-        const auto dataType = deserializeDataType(dataTypeStr);
+        // description is optional
+        auto description = std::string("");
+        try
+        {
+            description = modbusRegisterMappingJson.at("description").get<std::string>();
+        }
+        catch (std::exception&)
+        {
+        }    // it will be empty
```
[the remark this answers is not in the document]
Comment (nvuletic) [reply src/modbus/ModbusRegisterMapping.cpp]: That means that the already assigned value won't be changed. If the field is optional, there is no way to detect if it exists or not, other than to *try* and read it. I'm moving the assignment of values to catch, just to make it more readable (I tried to write it in a way where the compiler would be 100% sure that a value will exist).

File: src/modbus/ModbusClient.cpp [hunk 8]
```diff
@@ -370,36 +467,19 @@ bool ModbusClient::readInputRegisters(int address, int number, std::vector<float
...
-        std::uint8_t mask = 1;
-        for (int j = 0; j < bits; ++j)
-        {
-            std::uint8_t tmpValue = mask & tmpValues[i];
-            values.push_back(static_cast<bool>(tmpValue));
-            mask << 1;
-        }
+    for (int i = 0; i < bits_read; i++)
+    {
+        values.push_back(static_cast<bool>(tmpValues[i]));
```
[the remark this answers is not in the document]
Comment (nvuletic) [reply src/modbus/ModbusClient.cpp]: The logic that was here in place before didn't work. I copied this from readInputContacts, as they got this sorted out.

File: src/modbus/ModbusBridge.h [hunk 9]
```diff
@@ -57,33 +64,57 @@ class ModbusBridge : public ActuationHandlerPerDevice,
...
+
     void run();
 
     void readAndReportModbusRegistersValues();
 
     ModbusClient& m_modbusClient;
 
     int m_timeoutIterator;
-    const std::vector<int> m_timeoutDurations = {1, 5, 10, 15, 30, 60, 300, 600, 1800, 3600};
+    std::vector<int> m_timeoutDurations = {1, 5, 10, 15, 30, 60, 300, 600, 1800, 3600};
```
[the remark this answers is not in the document]
Comment (nvuletic) [reply src/modbus/ModbusBridge.h]: Rebase missed that out

File: src/modbus/ModbusRegisterMapping.cpp [hunk 10]
```diff
@@ -102,29 +154,133 @@ std::vector<wolkabout::ModbusRegisterMapping> ModbusRegisterMappingFactory::from
...
+                        gotAddress = true;
+                    }
+                    catch (std::exception&)
+                    {
+                    }
+                    try
+                    {
+                        auto tempLabelMap = std::make_shared<std::map<std::string, int>>(
+                          modbusRegisterMappingJson.at("labelMap").get<std::map<std::string, int>>());
+                        typedef std::function<bool(std::pair<std::string, int>, std::pair<std::string, int>)>
```
[the remark this answers is not in the document]
Comment (nvuletic) [reply src/modbus/ModbusRegisterMapping.cpp]: Resolved, but not in this snippet, isn't marked as outdated.

File: src/modbus/ModbusRegisterMapping.cpp [hunk 11]
```diff
@@ -102,29 +154,133 @@ std::vector<wolkabout::ModbusRegisterMapping> ModbusRegisterMappingFactory::from
...
+                          Comparator;
+                        Comparator compFunctor = [](std::pair<std::string, int> elem1,
+                                                    std::pair<std::string, int> elem2) {
+                            return elem1.second < elem2.second;
+                        };
+                        std::set<std::pair<std::string, int>, Comparator> setOfWords(tempLabelMap->begin(),
```
[the remark this answers is not in the document]
Comment (nvuletic) [reply src/modbus/ModbusRegisterMapping.cpp]: Actually found to be necessary

File: src/modbus/ModbusClient.cpp [hunk 12]
```diff
@@ -338,47 +336,37 @@ bool ModbusClient::writeHoldingRegister(int address, float value)
 
 bool ModbusClient::writeHoldingRegisters(int address, std::vector<short>& values)
 {
-    auto* arr = new uint16_t[values.size()];
+    auto arr = std::vector<uint16_t>();
     for (int i = 0; i < values.size(); i++)
     {
         ModbusValue modbusValue;
         modbusValue.signedShortValue = values[i];
         arr[i] = modbusValue.unsignedShortValue;
```
Comment (milanzlicic) [inline src/modbus/ModbusClient.cpp]: Indexing vector of size 0 will case segfault; create it with desired size first

**Pull request merged**